pkg/bucketrequest/bucketrequest.go

```diff
@@ -0,0 +1,174 @@
+	bucketRequest := obj
+	status, err := b.provisionBucketRequestOperation(ctx, bucketRequest)
+	if err == nil || status == "Finished" {
+		// Provisioning is 100% finished / not in progress.
+		switch err {
+		case nil:
+			glog.V(5).Infof("BucketRequest processing succeeded, removing bucketRequest %s from bucketRequests in progress", bucketRequest.UID)
+		case util.ErrStopProvision:
+			glog.V(5).Infof("Stop provisioning, removing bucketRequest %s from bucketRequests in progress", bucketRequest.UID)
+			// Our caller would requeue if we pass on this special error; return nil instead.
+			err = nil
+		default:
+			glog.V(2).Infof("Final error received, removing buckerRequest %s from bucketRequests in progress", bucketRequest.UID)
+		}
+		return err
+	}
+	if status == "InBackground" {
+		//nothing for now
+	} else {
+		// status == ProvisioningNoChange.
+		// Don't change bucketRequestsInProgress:
+		// - the bucketRequest is already there if previous status was ProvisioningInBackground.
+		// - the bucketRequest is not there if if previous status was "Finished".
+	}
+	return nil
+}
+
+// update processes any updates  made to the bucket request
+func (b *bucketRequestListener) Update(ctx context.Context, old, new *v1alpha1.BucketRequest) error {
+	glog.V(1).Infof("update called for bucket %v", old)
+	return nil
+}
+
+// Delete processes a bucket for which bucket request is deleted
+func (b *bucketRequestListener) Delete(ctx context.Context, obj *v1alpha1.BucketRequest) error {
+	return nil
+}
+
+// provisionBucketRequestOperation attempts to provision a bucket for the given bucketRequest.
+// Returns nil error only when the bucket was provisioned (in which case it also returns "Finished"),
+// a normal error when the bucket was not provisioned and provisioning should be retried (requeue the bucketRequest),
+// or the special errStopProvision when provisioning was impossible and no further attempts to provision should be tried.
+func (b *bucketRequestListener) provisionBucketRequestOperation(ctx context.Context, bucketRequest *v1alpha1.BucketRequest) (string, error) {
+	// Most code here is identical to that found in controller.go of kube's  controller...
+	bucketClassName := b.GetBucketClass(bucketRequest)
+
+	//  A previous doProvisionBucketRequest may just have finished while we were waiting for
+	//  the locks. Check that bucket (with deterministic name) hasn't been provisioned
+	//  yet.
+	bucket := b.FindBucket(ctx, bucketRequest)
+	if bucket != nil {
+		// bucket has been already provisioned, nothing to do.
+		glog.Info("Bucket already exists", bucket.Name)
+		return "Finished", util.ErrStopProvision
+	}
+
+	bucketClass, err := b.bucketClient.ObjectstorageV1alpha1().BucketClasses().Get(ctx, bucketClassName, metav1.GetOptions{})
+	if bucketClass == nil {
+		// bucketclass does not exist in order to create a bucket
+		return "InvalidBucketClass", util.ErrStopProvision
+	}
+
+	glog.Info(logOperation("created", "bucket %q provisioned", bucket))
```

Comment:
why is success declared here? The provisioner hasn't even had a chance to see the new bucket instance yet.

Reply:
Success is that the bucket is created but status is not available until provisioner acts on it.

Reply:
Good Question. @brahmaroutu

Reply:
Doesn't Created give the false impression that the bucket is created in the backend?

Reply:
> Success is that the bucket is created but status is not available until provisioner acts on it.

But the log msg states that the bucket was provisioned and this is not the case.